language/stdlib/modules/association.move

```diff
@@ -0,0 +1,112 @@
+// Implements logic for registering addresses as association addresses, and
+// determining if the sending account is an association account.
+// Errors:
+// 1000 -> INVALID_GENESIS_ADDR
+// 1001 -> INSUFFICIENT_PRIVILEGES
+// 1002 -> NOT_AN_ASSOCIATION_ACCOUNT
+// 1003 -> ACCOUNT_DOES_NOT_HAVE_PRIVILEGE
+// 1004 -> ACCOUNT_DOES_NOT_HAVE_PRIVILEGE_RESOURCE
+address 0x0:
+
+module Association {
+    use 0x0::Transaction;
+
+    // The root account privilege. This is created at genesis and has
+    // special privileges (e.g. removing an account as an association
+    // account). It cannot be removed.
+    resource struct Root { }
+
+    // There are certain association capabilities that are more
+    // privileged than other association operations. This resource with the
+    // type representing that privilege is published under the privileged
+    // account.
+    resource struct PrivilegedCapability<Privilege> { is_certified: bool }
+
+    // A type tag to mark that this account is an association account.
+    // It cannot be used for more specific/privileged operations.
+    struct T { }
+
+    // Initialization is called in genesis. It publishes the root resource
+    // under the root_address() address, marks it as a normal
+    // association account.
+    public fun initialize() {
+        let sender = Transaction::sender();
+        Transaction::assert(sender == root_address(), 1000);
+        move_to_sender(Root{ });
+        move_to_sender(PrivilegedCapability<T>{ is_certified: true });
+    }
+
+    // Publish a specific privilege under the sending account.
+    public fun apply_for_privilege<Privilege>() {
+        move_to_sender(PrivilegedCapability<Privilege>{ is_certified: false });
+    }
+
+    // Certify the privileged capability published under for_addr.
+    public fun grant_privilege<Privilege>(for_addr: address)
+    acquires PrivilegedCapability {
+        assert_sender_is_root();
+        Transaction::assert(exists<PrivilegedCapability<Privilege>>(for_addr), 1003);
+        borrow_global_mut<PrivilegedCapability<Privilege>>(for_addr).is_certified = true;
+    }
+
+    // Return whether the `addr` has the specified `Privilege`.
+    public fun has_privilege<Privilege>(addr: address): bool
+    acquires PrivilegedCapability {
+        addr_is_association(addr) &&
+        exists<PrivilegedCapability<Privilege>>(addr) &&
+        borrow_global<PrivilegedCapability<Privilege>>(addr).is_certified
+    }
+
+    // Remove the `Privilege` from the address at `addr`. The sender must
+    // be the root association account. The `Privilege` need not be
+    // certified.
+    public fun remove_privilege<Privilege>(addr: address)
```

Comment:
One use-case I'm wondering about (for both certification and removal): what if the holder of a `PrivilegedCapability` wants to self-impose additional constraints on its use (e.g.,  adding new validators at most once a month)? The easy way to do that is something like `resource struct RateLimitedAddValidator { cap: PrivilegedCapability<AddValidator::T>, last_addition: u64 }`. But then certifying/revoking becomes more complicated... 

In any case, this isn't a critica; problem, but wanted to bring it up. It might be reasonable to do something simple like: 
- "A `PrivilegedCapability` must always be published at the top-level + certified first, then can be moved into a wrapper like `RateLimitedAddValidator` afterward"
- We could provide an alternative API for using `Offer::T` to give an address a pre-approved `PrivilegedCapability` to be published anywhere (including in a wrapper)
- We could choose to use account freezing to implement `remove_privilege`, as you suggested before. It's kind of difficult to imagine a case where you'd want to forcibly revoke one privilege but leave others + the basic account functionality intact.

Reply:
Good question. There are three main reasons--both of them more design-related--why I would argue that we may not want this ability of storing `PrivilegedCapabilities` within other structures:

1. The declaring module (usually) makes their own privilege (in your example, the `ValidatorAdmission` module would declare the `AddValidator` struct cf. the example below). Particularly, I view the addition of constraints on a privilege as being a module specific property, whereas the `PrivilegeCapability` is a property of the account/association that ensures only accounts holding specific privileges can initiate specific operations.
2. My second concern is about allowing a `PrivilegedCapability` to be "free'd" in the  manner you described. Once the privilege is held by another module we don't necessarily have any control over the logic relating to the relinquishing of that privilege, and therefore any future owners of the `PrivilegedCapability`. In particular, if we want to say that adding a validator requires 32-32 multisig account, this is much easier to enforce if the resource is required to be published at the top-level and therefore under a specific account.
3. My third concern is that there could be a privilege that we no longer want to exist. However, if these `PrivilegedCapability` resources are not published at the top-level the `Association` module has no way of decertifying it or removing it. And, since we don't have (easy) control of where it could end up (see 2.), this could lead to us having an errant privilege that we no longer want, but have no (easy) way of removing from the system.

To provide an example of how I would envision rate-limited validator admission, I feel as though this could also be naturally encoded as:
```rust
module ValidatorAdmission {
        use 0x0::Association;
        resource struct RateLimitedAddValidator { last_addition: u64 }
        struct AddValidator { }

        ...

        fun add_validator(...) {
        	Association::has_privilege<AddValidator>(sender());
        	let rate_limit = borrow_global<RateLimitedValidator>(..);
            // logic
        }
}
```

Reply:
All of your points are good ones (and I don't mean for them to block this PR). 

I agree that it's better to add constraints *inside* the relevant module if 
A. you know what they are going to be in advance and 
B. want all invocations of the operation to be subject to them

(both of which are true for my example). But I think there are also cases where you suspect that the constraints will change in the future, and you don't need all invocations of the operation to be subject to them (but want to allow individual privilege holders to place additional constraints on themselves).

The specific example I am struggling with how to do is `ConfigHash` equality checking (to support AOS). Conceptually, this is an additional constraint that each holder of a (e.g.) `PrivilegedCapability<AddValidator::T>` wants to impose on themselves. It has nothing to do with the privilege checking for `AddValidator`; some folks with the capability may not want (or be able) to satisfy that constraint.

The natural way to do this is letting the capability float "free" (as you describe in (2)) and then locked in a wrapper. But I totally agree with your concerns about revocation in the general case of this problem (in this specific case, I think we could probably revoke via freezing and that would be fine). It would be great if there was a design pattern that let the capability always be stuck at the top level (preserving revocability), but allowed the holder to add arbitrary additional constraints on actually using the capability...